File: Easy/maximum_depth_of_binary_tree/maximum_depth_of_binary_tree.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    public:
        int maxDepth(TreeNode* root) {
            if (root == nullptr) {
                return 0;
            }
            int leftDepth = maxDepth(root->left);
            int rightDepth = maxDepth(root->right);
            return 1 + max(leftDepth, rightDepth);
        }
};

// Example usage:
int main() {
    Solution solution1;
    TreeNode* root1 = new TreeNode(3);
    root1->left = new TreeNode(9);
    root1->right = new TreeNode(20);
    root1->right->left = new TreeNode(15);
    root1->right->right = new TreeNode(7);
    int result1 = solution1.maxDepth(root1);
    cout << "Input tree: [3,9,20,null,null,15,7]" << endl;
    cout << "Maximum depth of the first tree: " << result1 << endl;

    Solution solution2;
    TreeNode* root2 = new TreeNode(1);
    root2->left = new TreeNode(2);
    int result2 = solution2.maxDepth(root2);
    cout << endl;
    cout << "Input tree: [1,null,2]" << endl;
    cout << "Maximum depth of the second tree: " << result2 << endl;
    return 0;
}

File: Easy/maximum_depth_of_binary_tree/maximum_depth_of_binary_tree.py
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def maxDepth(root: Optional[TreeNode]) -> int:
    if root is None:
        return 0
    else:
        return 1 + max(maxDepth(root.left), maxDepth(root.right)) #Doing this recursively to find the maximum depth of the binary tree.

if __name__ == "__main__":
    print("Testing maxDepth function:")
    # Example tree: [3, 9, 20, null, null, 15, 7]
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    print("For the input [3, 9, 20, null, null, 15, 7], the output is:", maxDepth(root)) #Output: 3

    # Example tree: [1, null, 2]
    root2 = TreeNode(1)
    root2.right = TreeNode(2)
    print("For the input [1, null, 2], the output is:", maxDepth(root2)) #Output: 2
